File: app/src/main/java/com/example/firefighters/repositories/EmergencyRepository.java
package com.example.firefighters.repositories;

import com.example.firefighters.models.EmergencyModel;
import com.example.firefighters.tools.ConstantsValues;
import com.example.firefighters.tools.FirebaseManager;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.EventListener;
import com.google.firebase.firestore.FirebaseFirestoreException;
import com.google.firebase.firestore.Query;
import com.google.firebase.firestore.QuerySnapshot;

import org.jetbrains.annotations.NotNull;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

public class EmergencyRepository {

    private static EmergencyRepository instance;

    public static EmergencyRepository getInstance() {
        if (instance == null)
            instance = new EmergencyRepository();
        return instance;
    }

    public LiveData<QuerySnapshot> getEmergenciesQuery(DocumentSnapshot lastDocument, String lastFilter, Query.Direction lastOrder, int limitCount) {
        if (lastFilter == null || lastFilter.isEmpty())
            lastFilter = ConstantsValues.FILTER_NAME;
        if (lastOrder == null)
            lastOrder = Query.Direction.DESCENDING;
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        if (lastDocument != null){
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                    .orderBy(lastFilter, lastOrder)
                    .startAfter(lastDocument)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }else{
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                    .orderBy(lastFilter, lastOrder)
                    .limit(limitCount)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            data.setValue(task.getResult());
                        }
                    });
        }
        return data;
    }
    public LiveData<QuerySnapshot> getEmergenciesQuerySnapshot(String lastFilter, Query.Direction lastOrder) {
        if (lastFilter == null || lastFilter.isEmpty())
            lastFilter = ConstantsValues.FILTER_NAME;
        if (lastOrder == null)
            lastOrder = Query.Direction.DESCENDING;
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                .orderBy(lastFilter, lastOrder)
                .addSnapshotListener(new EventListener<QuerySnapshot>() {
                    @Override
                    public void onEvent(@Nullable @org.jetbrains.annotations.Nullable QuerySnapshot value, @Nullable @org.jetbrains.annotations.Nullable FirebaseFirestoreException error) {
                        data.setValue(value);
                    }
                });
        return data;
    }

    public LiveData<EmergencyModel> getEmergencyWorkingOnModel(String currentUnit) {
        MutableLiveData<EmergencyModel> data = new MutableLiveData<>();
        if (currentUnit == null || currentUnit.isEmpty()) {
            data.setValue(null);
        } else {
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                    .whereEqualTo("status", ConstantsValues.WORKING)
                    .whereEqualTo("currentUnit", currentUnit)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                        @Override
                        public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                            if (task.isSuccessful()){
                                if (task.getResult().getDocuments().size() > 0){
                                    EmergencyModel emergencyModel = task.getResult().getDocuments().get(0).toObject(EmergencyModel.class);
                                    data.setValue(emergencyModel);
                                }else{
                                    data.setValue(null);
                                }
                            }else{
                                data.setValue(null);
                            }
                        }
                    });
        }
        return data;
    }

    public LiveData<QuerySnapshot> getEmergencyWorkingOn(String currentUnit) {
        MutableLiveData<QuerySnapshot> data = new MutableLiveData<>();
        if (currentUnit == null || currentUnit.isEmpty()) {
            data.setValue(null);
        } else {
            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                    .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                    .whereEqualTo("currentUnit", currentUnit)
                    .whereEqualTo("status", ConstantsValues.WORKING)
                    .addSnapshotListener(new EventListener<QuerySnapshot>() {
                        @Override
                        public void onEvent(@Nullable @org.jetbrains.annotations.Nullable QuerySnapshot value, @Nullable @org.jetbrains.annotations.Nullable FirebaseFirestoreException error) {
                            data.setValue(value);
                        }
                    });
        }
        return data;
    }

    public LiveData<Integer> saveEmergency(EmergencyModel emergencyModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                .orderBy("id", Query.Direction.DESCENDING)
                .limit(1)
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            long lastId = 0;
                            if (task.getResult().size() > 0) {
                                EmergencyModel em = task.getResult().getDocuments().get(0).toObject(EmergencyModel.class);
                                if (em != null)
                                    lastId = em.getId();
                            }
                            emergencyModel.setId(lastId+1);
                            long finalLastId = lastId+1;
                            FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                                    .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                                    .add(emergencyModel)
                                    .addOnCompleteListener(new OnCompleteListener<DocumentReference>() {
                                        @Override
                                        public void onComplete(@NonNull @NotNull Task<DocumentReference> task) {
                                            if (task.isSuccessful()){
                                                data.setValue((int) finalLastId);
                                            }else{
                                                data.setValue(-1);
                                            }
                                        }
                                    });
                        }else {
                            data.setValue(-1);
                        }
                    }
                });

        return data;
    }
    public LiveData<Integer> deleteEmergency(EmergencyModel emergencyModel) {
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                .whereEqualTo("id", emergencyModel.getId())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            for (DocumentSnapshot document: task.getResult()) {
                                document.getReference()
                                        .delete()
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull @NotNull Task<Void> task) {
                                                if(task.isSuccessful()){
                                                    data.setValue(1);
                                                }else{
                                                    data.setValue(-1);
                                                }
                                            }
                                        });
                            }
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }
    public LiveData<Integer> updateEmergency(EmergencyModel emergencyModel){
        MutableLiveData<Integer> data = new MutableLiveData<>();
        FirebaseManager.getInstance().getFirebaseFirestoreInstance()
                .collection(ConstantsValues.EMERGENCIES_COLLECTION)
                .whereEqualTo("id", emergencyModel.getId())
                .get()
                .addOnCompleteListener(new OnCompleteListener<QuerySnapshot>() {
                    @Override
                    public void onComplete(@NonNull @NotNull Task<QuerySnapshot> task) {
                        if (task.isSuccessful()){
                            for (DocumentSnapshot document: task.getResult()) {
                                document.getReference()
                                        .update("senderMail", emergencyModel.getSenderMail(),
                                                "messageId", emergencyModel.getMessageId(),
                                                "longitude", emergencyModel.getLongitude(),
                                                "latitude", emergencyModel.getLatitude(),
                                                "gravity", emergencyModel.getGravity(),
                                                "status", emergencyModel.getStatus(),
                                                "currentUnit", emergencyModel.getCurrentUnit()
                                                )
                                        .addOnCompleteListener(new OnCompleteListener<Void>() {
                                            @Override
                                            public void onComplete(@NonNull @NotNull Task<Void> task) {
                                                if(task.isSuccessful()){
                                                    data.setValue(1);
                                                }else{
                                                    data.setValue(-1);
                                                }
                                            }
                                        });
                            }
                        }else{
                            data.setValue(-1);
                        }
                    }
                });
        return data;
    }
}

File: app/src/main/java/com/example/firefighters/repositories/EmergencyRepository.kt
package com.example.firefighters.repositories

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.example.firefighters.models.EmergencyModel
import com.example.firefighters.utils.ConstantsValues
import com.example.firefighters.utils.FirebaseUtils
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.Query
import com.google.firebase.firestore.QuerySnapshot

class EmergencyRepository {
    fun getEmergenciesQuery(
        lastDocument: DocumentSnapshot?,
        lastFilter: String?,
        lastOrder: Query.Direction?,
        limitCount: Int
    ): LiveData<QuerySnapshot> {
        val tempLastFilter: String = lastFilter ?: ConstantsValues.FILTER_NAME
        val tempLastOrder: Query.Direction = lastOrder ?: Query.Direction.DESCENDING
        val data = MutableLiveData<QuerySnapshot>()
        if (lastDocument != null) {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
                ?.orderBy(tempLastFilter, tempLastOrder)
                ?.startAfter(lastDocument)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
                ?.orderBy(tempLastFilter, tempLastOrder)
                ?.limit(limitCount.toLong())
                ?.get()
                ?.addOnCompleteListener { task -> data.value = task.result }
        }
        return data
    }

    fun getEmergenciesQuerySnapshot(
        lastFilter: String?,
        lastOrder: Query.Direction?
    ): LiveData<QuerySnapshot?> {
        val tempLastFilter: String = lastFilter ?: ConstantsValues.FILTER_NAME
        val tempLastOrder: Query.Direction = lastOrder ?: Query.Direction.DESCENDING
        val data = MutableLiveData<QuerySnapshot?>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.orderBy(tempLastFilter, tempLastOrder)
            ?.addSnapshotListener { value, _ -> data.value = value }
        return data
    }

    fun getLastQuerySnapshot(
        lastFilter: String?,
        lastOrder: Query.Direction?
    ): LiveData<QuerySnapshot?> {
        val tempLastFilter: String = lastFilter ?: ConstantsValues.FILTER_NAME
        val tempLastOrder: Query.Direction = lastOrder ?: Query.Direction.DESCENDING
        val data = MutableLiveData<QuerySnapshot?>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.orderBy(tempLastFilter, tempLastOrder)
            ?.limit(1)
            ?.addSnapshotListener { value, _ -> data.value = value }
        return data
    }

    fun getEmergencyWorkingOnModel(currentUnit: String?): LiveData<EmergencyModel?> {
        val data = MutableLiveData<EmergencyModel?>()
        if (currentUnit.isNullOrEmpty()) {
            data.setValue(null)
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
                ?.whereEqualTo("status", ConstantsValues.WORKING)
                ?.whereEqualTo("currentUnit", currentUnit)
                ?.get()
                ?.addOnCompleteListener { task ->
                    if (task.isSuccessful) {
                        if ((task.result?.documents?.size ?: 0) > 0) {
                            val emergencyModel = task.result?.documents?.get(0)?.toObject(
                                EmergencyModel::class.java
                            )
                            data.setValue(emergencyModel)
                        } else {
                            data.setValue(null)
                        }
                    } else {
                        data.setValue(null)
                    }
                }
        }
        return data
    }

    fun getEmergencyWorkingOn(currentUnit: String?): LiveData<QuerySnapshot?> {
        val data = MutableLiveData<QuerySnapshot?>()
        if (currentUnit.isNullOrEmpty()) {
            data.setValue(null)
        } else {
            FirebaseUtils.instance?.firebaseFirestoreInstance
                ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
                ?.whereEqualTo("currentUnit", currentUnit)
                ?.whereEqualTo("status", ConstantsValues.WORKING)
                ?.addSnapshotListener { value, _ -> data.value = value }
        }
        return data
    }

    fun saveEmergency(emergencyModel: EmergencyModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.orderBy("id", Query.Direction.DESCENDING)
            ?.limit(1)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    var lastId: Long = 0
                    if ((task.result?.size() ?: 0) > 0) {
                        val em = task.result?.documents?.get(0)?.toObject(
                            EmergencyModel::class.java
                        )
                        if (em != null) lastId = em.id
                    }
                    emergencyModel.id = lastId + 1
                    val finalLastId = lastId + 1
                    FirebaseUtils.instance?.firebaseFirestoreInstance
                        ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
                        ?.add(emergencyModel)
                        ?.addOnCompleteListener {
                            if (it.isSuccessful) {
                                data.setValue(finalLastId.toInt())
                            } else {
                                data.setValue(-1)
                            }
                        }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun deleteEmergency(emergencyModel: EmergencyModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.whereEqualTo("id", emergencyModel.id)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result ?: listOf()) {
                        document.reference
                            .delete()
                            .addOnCompleteListener {
                                if (it.isSuccessful) {
                                    data.setValue(1)
                                } else {
                                    data.setValue(-1)
                                }
                            }
                    }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }

    fun updateEmergency(emergencyModel: EmergencyModel): LiveData<Int> {
        val data = MutableLiveData<Int>()
        FirebaseUtils.instance?.firebaseFirestoreInstance
            ?.collection(ConstantsValues.EMERGENCIES_COLLECTION)
            ?.whereEqualTo("id", emergencyModel.id)
            ?.get()
            ?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    for (document in task.result ?: listOf()) {
                        document.reference
                            .update(
                                "senderMail", emergencyModel.senderMail,
                                "messageId", emergencyModel.messageId,
                                "longitude", emergencyModel.longitude,
                                "latitude", emergencyModel.latitude,
                                "gravity", emergencyModel.gravity,
                                "status", emergencyModel.status,
                                "currentUnit", emergencyModel.currentUnit
                            )
                            .addOnCompleteListener {
                                if (it.isSuccessful) {
                                    data.setValue(1)
                                } else {
                                    data.setValue(-1)
                                }
                            }
                    }
                } else {
                    data.setValue(-1)
                }
            }
        return data
    }
}
